nemo_aligner/models/nlp/gpt/gpt_steerlm_model.py

```diff
@@ -0,0 +1,357 @@
+    }
+    if "weight" in data:
+        batch.update({"weight": data["weight"]})
+    if "avg_num_valid_tokens_in_ub" in data:
+        batch.update({"avg_num_valid_tokens_in_ub": data["avg_num_valid_tokens_in_ub"]})
+    return batch
+
+
+class GPTSteerLMModel(GPTSFTModel):
+    def loss_func(self, loss_mask, output_tensor):
+        # loss is the nll loss for each token in the batch
+        losses = output_tensor.float()
+        # losses [batch_size, sequence_length]
+        # calculate sum of log(p) for all tokens in a one batch item
+        # this will be later used to calculate the baseline weights
+        loss = torch.sum(losses * loss_mask, dim=-1)
+        # if parallel_state.get_context_parallel_world_size() > 1:
+        #     torch.distributed.all_reduce(loss, group=parallel_state.get_context_parallel_group())
+        return loss
+
+    def weight_loss_func(self, loss_mask, avg_num_valid_tokens_in_ub, output_tensor, weight):
+        # loss is the nll loss for each token in the batch
+        losses = output_tensor.float()
+        loss_mask = loss_mask.float()
+        # apply the per batch weight
+        losses = losses * weight[:, None]
+        # losses [batch_size, sequence_length]
+        # loss per batch item, log(p) for all tokens in one batch item
+        loss = torch.sum(losses * loss_mask, dim=-1)
+        # average lop(p) by the number of valid tokens in the microbatch and scale it by the (number of microbatches / batch)
+        loss = loss / avg_num_valid_tokens_in_ub
+        # sum log(p) in microbatch
+        loss = loss.sum()
+        if parallel_state.get_context_parallel_world_size() > 1:
+            torch.distributed.all_reduce(loss, group=parallel_state.get_context_parallel_group())
+        return loss
+
+    def get_forward_output_only_for_weight_func(self):
+        def fwd_output_only_func(dataloader_iter, model):
+            with torch.no_grad():
+                batch = next(dataloader_iter)
+                batch = {key: val.cuda(non_blocking=True) for key, val in batch.items()}
+
+                forward_args = {
+                    "input_ids": batch["tokens"],
+                    "position_ids": batch["position_ids"],
+                    "attention_mask": None if self.get_attention_mask_from_fusion else batch["attention_mask"],
+                    "labels": batch["labels"],
+                }
+
+                output_tensor = model(**forward_args)
+
+                def id_func(output_tensor):
+                    nll = self.loss_func(batch["loss_mask"], output_tensor)
+                    return output_tensor, {"loss": nll}
+
+            return output_tensor, id_func
+
+        return fwd_output_only_func
+
+    def get_forward_output_and_loss_func(self, validation_step=False, tuning=False):
+        def fwd_output_and_loss_func(dataloader_iter, model, checkpoint_activations_all_layers=None):
+
+            # Get data batch
+            batch = get_batch(dataloader_iter, tuning)
```

Comment:
why do you need this function? can't you just directly use the batch here? also it seems like the tuning argument is not used